content/en/blog/_posts/2023-08-21-job-update-post.md

```diff
@@ -0,0 +1,231 @@
+title: "Kubernetes 1.28: Improved failure handling for Jobs"
+date: 2023-08-21
+slug: kubernetes-1-28-jobapi-update
+---
+
+**Authors:** Kevin Hannon (G-Research), Michał Woźniak (Google)
+
+This blog discusses two new features in Kubernetes 1.28 to improve Jobs for batch
+users: [Pod replacement policy](/docs/concepts/workloads/controllers/job/#pod-replacement-policy)
+and [Backoff limit per index](/docs/concepts/workloads/controllers/job/#backoff-limit-per-index).
+
+These features continue the effort started by the
+[Pod failure policy](/docs/concepts/workloads/controllers/job/#pod-failure-policy)
+to improve the handling of Pod failures in a Job.
+
+## Pod replacement policy {#pod-replacement-policy}
+
+By default, when a pod enters a terminating state (e.g. due to preemption or
+eviction), Kubernetes immediately creates a replacement Pod. Therefore, both Pods are running
+at the same time. In API terms, a pod is considered terminating when it has a
+`deletionTimestamp` and it has a phase `Pending` or `Running`.
+
+The scenario when two Pods are running at a given time is problematic for
+some popular machine learning frameworks, such as
+TensorFlow and [JAX](https://jax.readthedocs.io/en/latest/), which require at most one Pod running at the same time,
+for a given index. 
+Tensorflow gives the following error if two pods are running for a given index.
+
+```
+ /job:worker/task:4: Duplicate task registration with task_name=/job:worker/replica:0/task:4
+```
+
+See more details in the ([issue](https://github.com/kubernetes/kubernetes/issues/115844)).
+
+
+Creating the replacement Pod before the previous one fully terminates can also
+cause problems in clusters with scarce resources or with tight budgets, such as:
+* cluster resources can be difficult to obtain for Pods pending to be scheduled,
+  as Kubernetes might take a long time to find available nodes until the existing
+  Pods are fully terminated.
+* if cluster autoscaler is enabled, the replacement Pods might produce undesired
+  scale ups.
+
+### How can you use it? {#pod-replacement-policy-how-to-use}
+
+This is an alpha feature, which you can enable by turning on `JobPodReplacementPolicy`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) in
+your cluster.
+
+Once the feature is enabled in your cluster, you can use it by creating a new Job that specifies a
+`podReplacementPolicy` field as shown here:
+
+```yaml
+kind: Job
+metadata:
+  name: new
+  ...
+spec:
+  podReplacementPolicy: Failed
+  ...
+```
+
+In that Job, the Pods would only be replaced once they reached the `Failed` phase,
+and not when they are terminating.
+
+Additionally, you can inspect the `.status.terminating` field of a Job. The value
+of the field is the number of Pods owned by the Job that are currently terminating.
+
+```shell
+kubectl get jobs/myjob -o=jsonpath='{.items[*].status.terminating}'
+```
+
+```
+3 # three Pods are terminating and have not yet reached the Failed phase
```

Comment:
The actual output doesn't include a comment. I hope this is obvious to all readers.